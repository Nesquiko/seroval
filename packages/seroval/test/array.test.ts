import { describe, it, expect } from 'vitest';
import {
  AsyncServerValue,
  deserialize,
  fromJSON,
  serialize,
  serializeAsync,
  ServerValue,
  toJSONAsync,
  toJSON,
} from '../src';

describe('arrays', () => {
  describe('serialize', () => {
    it('supports Arrays', () => {
      const example = [1, 2, 3];
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<number[]>(result);
      expect(back).toBeInstanceOf(Array);
      expect(back[0]).toBe(example[0]);
      expect(back[1]).toBe(example[1]);
      expect(back[2]).toBe(example[2]);
    });
    it('supports self recursion', () => {
      const example: ServerValue[] = [];
      example[0] = example;
      example[1] = example;
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<ServerValue[]>(result);
      expect(back[0]).toBe(back);
      expect(back[1]).toBe(back);
    });
  });
  describe('serializeAsync', () => {
    it('supports Arrays', async () => {
      const example = [1, 2, 3];
      const result = await serializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<number[]>>(result);
      expect(back).toBeInstanceOf(Array);
      expect(back[0]).toBe(example[0]);
      expect(back[1]).toBe(example[1]);
      expect(back[2]).toBe(example[2]);
    });
    it('supports self recursion', async () => {
      const example: AsyncServerValue[] = [];
      example[0] = Promise.resolve(example);
      example[1] = Promise.resolve(example);
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof example>(result);
      expect(await back[0]).toBe(back);
      expect(await back[1]).toBe(back);
    });
  });
  describe('toJSON', () => {
    it('supports Arrays', () => {
      const example = [1, 2, 3];
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<number[]>(result);
      expect(back).toBeInstanceOf(Array);
      expect(back[0]).toBe(example[0]);
      expect(back[1]).toBe(example[1]);
      expect(back[2]).toBe(example[2]);
    });
    it('supports self recursion', () => {
      const example: ServerValue[] = [];
      example[0] = example;
      example[1] = example;
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<ServerValue[]>(result);
      expect(back[0]).toBe(back);
      expect(back[1]).toBe(back);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Arrays', async () => {
      const example = [1, 2, 3];
      const result = await toJSONAsync(Promise.resolve(example));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<number[]>>(result);
      expect(back).toBeInstanceOf(Array);
      expect(back[0]).toBe(example[0]);
      expect(back[1]).toBe(example[1]);
      expect(back[2]).toBe(example[2]);
    });
    it('supports self recursion', async () => {
      const example: AsyncServerValue[] = [];
      example[0] = Promise.resolve(example);
      example[1] = Promise.resolve(example);
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof example>(result);
      expect(await back[0]).toBe(back);
      expect(await back[1]).toBe(back);
    });
  });
});
describe('array holes', () => {
  describe('serialize', () => {
    it('supports array holes', () => {
      const example = new Array(10);
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<undefined[]>(result);
      expect(0 in back).toBeFalsy();
      expect(back[0]).toBe(undefined);
      expect(back.length).toBe(example.length);
    });
  });
  describe('serializeAsync', () => {
    it('supports array holes', async () => {
      const example = new Array(10);
      const result = await serializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<undefined[]>>(result);
      expect(0 in back).toBeFalsy();
      expect(back[0]).toBe(undefined);
      expect(back.length).toBe(example.length);
    });
  });
  describe('toJSON', () => {
    it('supports array holes', () => {
      const example = new Array(10);
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<undefined[]>(result);
      expect(0 in back).toBeFalsy();
      expect(back[0]).toBe(undefined);
      expect(back.length).toBe(example.length);
    });
  });
  describe('toJSONAsync', () => {
    it('supports array holes', async () => {
      const example = new Array(10);
      const result = await toJSONAsync(Promise.resolve(example));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<undefined[]>>(result);
      expect(0 in back).toBeFalsy();
      expect(back[0]).toBe(undefined);
      expect(back.length).toBe(example.length);
    });
  });
});
