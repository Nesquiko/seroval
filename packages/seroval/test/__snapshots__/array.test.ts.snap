// Vitest Snapshot v1

exports[`array holes > serialize > supports array holes 1`] = `"[,,,,,,,,,,]"`;

exports[`array holes > serializeAsync > supports array holes 1`] = `"Promise.resolve([,,,,,,,,,,])"`;

exports[`array holes > toJSON > supports array holes 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":10,\\"a\\":[null,null,null,null,null,null,null,null,null,null]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`array holes > toJSONAsync > supports array holes 1`] = `"{\\"t\\":{\\"t\\":18,\\"i\\":0,\\"f\\":{\\"t\\":15,\\"i\\":1,\\"l\\":10,\\"a\\":[null,null,null,null,null,null,null,null,null,null]}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`arrays > serialize > supports Arrays 1`] = `"[1,2,3]"`;

exports[`arrays > serialize > supports self recursion 1`] = `"(h=>(h=[,,],h[1]=h[0]=h,h))()"`;

exports[`arrays > serializeAsync > supports Arrays 1`] = `"Promise.resolve([1,2,3])"`;

exports[`arrays > serializeAsync > supports self recursion 1`] = `"(h=>(h=[Promise.resolve().then(()=>h),Promise.resolve().then(()=>h)]))()"`;

exports[`arrays > toJSON > supports Arrays 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":3,\\"a\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":3}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`arrays > toJSON > supports self recursion 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":10,\\"i\\":0},{\\"t\\":10,\\"i\\":0}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[0]}"`;

exports[`arrays > toJSONAsync > supports Arrays 1`] = `"{\\"t\\":{\\"t\\":18,\\"i\\":0,\\"f\\":{\\"t\\":15,\\"i\\":1,\\"l\\":3,\\"a\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":3}]}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`arrays > toJSONAsync > supports self recursion 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":18,\\"i\\":1,\\"f\\":{\\"t\\":10,\\"i\\":0}},{\\"t\\":18,\\"i\\":2,\\"f\\":{\\"t\\":10,\\"i\\":0}}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[0]}"`;
